tools/cr/commit-msg.py

```diff
@@ -0,0 +1,151 @@
+# insertions.
+#
+# To install this script as a commit-msg hook, run the following command:
+#   cp tools/cr/commit-msg.py .git/hooks/commit-msg
+#
+
+import os
+import sys
+import subprocess
+import re
+
+PREFIXES_FOR_UPGRADE_COMMITS = [
+    'Update from Chromium ',
+    'Conflict-resolved patches from Chromium ',
+    'Update patches from Chromium ',
+    'Updated strings for Chromium ',
+]
+
+
+def main():
+    if os.getenv("GIT_REBASE"):
+        # There should be no changes to commits messages during a rebase.
+        return 0
+
+    commit_msg_file = sys.argv[1]
+
+    # Read the message provided so we can check for duplicate insertions
+    # beforehand.
+    with open(commit_msg_file, 'r') as f:
+        commit_message = f.read()
+
+    if not commit_message or not commit_message.split('\n', 1)[0].strip():
+        # We should not commmit if the commit message first line is empty as
+        # this will cause any tags to become the effective message.
+        return 1
+
+    # Skip fixup! changes.
+    if commit_message.startswith('fixup!'):
+        return 0
+
+    first_line = commit_message.split('\n', 1)[0]
+
+    # Getting the branch name to see if there's any tags that can be deduced
+    # from it.
+    branch_name = subprocess.check_output(
+        ['git', 'rev-parse', '--abbrev-ref', 'HEAD']).strip().decode()
+
+    tags = set()
+    # Checking for chromium upgrade branches
+    cr_branch_pattern = r"^cr\d+$"
+    if re.fullmatch(cr_branch_pattern, branch_name):
+        # The branch tag is excluded from the upgrade commit messages.
+        if not any(suffix in first_line
+                   for suffix in PREFIXES_FOR_UPGRADE_COMMITS):
+            tags.add(branch_name)
+    elif '+' in branch_name:
+        # This is the simple case where `canary+fix-failure` produces a
+        # [canary] tag.
+        tags.update(branch_name.split('+')[:-1])
+
+    # Checking for tags passed in via the environment variable TAGS
+    tags_env = os.getenv("TAGS")
+    if tags_env:
+        tags += tags_env.split(',')
+
+    # Making sure we are not re-inserting tags that are already in the commit
+    # message
+    tags = [f'[{tag}]' for tag in tags if f'[{tag}]' not in first_line]
+
+    # These are the report links that are added to the commit message when
+    # there's an upstream culprit.
+    culprit_hashes = []
+    if os.getenv("CULPRIT"):
+        culprit_hashes = os.getenv("CULPRIT").split(',')
+    culprit_hashes = [
+        culprit_hash for culprit_hash in culprit_hashes
+        if culprit_hash not in commit_message
+    ]
+
+    culprit_output = []
+    for culprit_hash in culprit_hashes:
+        culprit_output.append(
+            subprocess.check_output(
+                ['git', '-C', '../', 'log', '-1',
+                 culprit_hash]).decode().strip())
+
+    if tags:
+        commit_message = f'{" ".join(tags)} {commit_message}'
+
+    issues_env = os.getenv("ISSUE")
```

Comment:
should we get this from some kind of standardized branch naming convention instead?

Reply:
This is passed in manually, as in `ISSUE=45015 git commit`

Reply:
understood, but as an alternative could we also look at the branch name and have some convention for it? I have sometimes used `issues/[issue_number]`. So first check env, then branch name? The env doesn't work well for vscode.

Reply:
I'm already doing this convention thing with the branch name for the tag (e.g. `canary+` for a canary tag) so anything behind a `+` becomes a tag.

Reply:
I guess it could be -issue-# in the branch name.

Reply:
Fixed. If `issue-####` occurs in the branch name, that gets added to the commit message as a `Resolves` link